File: java/src/main/java/exercise/luhn/LuhnsAlgorithmCreditCardValidator.java
package exercise.luhn;


import java.util.List;
import java.util.stream.IntStream;

import static exercise.luhn.LuhnsAlgorithmCreditCardValidator.Functions.*;
import static java.util.Collections.reverse;
import static java.util.stream.Collectors.toList;

public class LuhnsAlgorithmCreditCardValidator implements CreditCardNumberValidator {

	public Boolean validate(String number) throws CreditCardNumberValidationException {
		List<Integer> doubled = reverseAndDoubleEverySecondDigit(toLong(number));
		int sumOfDigits = sumOfDigits(doubled);
		return isDivisibleByTenExactly(sumOfDigits);
	}

	private long toLong(String number) throws CreditCardNumberValidationException {
		try {
            String stripped = number.replace(" ", "");
            if (stripped.length() != 11) throw new CreditCardNumberValidationException("Credit card numbers must contain 11 digits");
			return Long.parseLong(stripped);
		} catch (NumberFormatException e) {
			throw new CreditCardNumberValidationException("Credit card numbers must contain only numbers");
		}
	}


	static class Functions {

		static List<Integer> reverseAndDoubleEverySecondDigit(Long number) {
			List<Integer> values = separateDigits(number);
			reverse(values);
			List<Integer> doubled = IntStream
					.range(0, values.size())
					.map(index -> isEven(index) ? values.get(index) : values.get(index) * 2)
					.boxed()
					.collect(toList());
			return doubled;
		}

        private static boolean isEven(int index) {
            return index % 2 == 0;
        }

        static List<Integer> separateDigits(Long number) {
            return number.toString()
					.chars()
					.map(Character::getNumericValue)
					.boxed()
					.collect(toList());
		}

		static Integer sumOfDigits(List<Integer> digits) {
			return digits
					.stream()
					.flatMapToInt(x -> separateDigits(Integer.toUnsignedLong(x)).stream().mapToInt(Integer::intValue))
					.sum();
		}


		static Boolean isDivisibleByTenExactly(Integer dividend) {
			if (dividend < 0) return false;
			return dividend % 10 == 0;
		}
	}
}

File: src/main/scala/exercise/luhn/scala/LuhnsAlgorithmCreditCardValidator.scala
package exercise.luhn.scala

import exercise.luhn.scala.LuhnsAlgorithmCreditCardValidator._

import scala.util.Try

class LuhnsAlgorithmCreditCardValidator extends CreditCardNumberValidator {

  def validate(number: String): Either[ValidationError, Boolean] = {
    for {
      candidate <- toLong(number)
      doubled   <- Right(reverseAndDoubleEverySecondDigit(candidate))
      sum       <- Right(sumOfDigits(doubled))
    } yield isDivisibleByTenExactly(sum)
  }
}

object LuhnsAlgorithmCreditCardValidator {
  def reverseAndDoubleEverySecondDigit(number: Long): List[Int] = {
    separateDigits(number)
      .reverse
      .zipWithIndex
      .map {
        case (value, index) if isEven(index) => value
        case (value, _)                      => value * 2
      }
  }

  private def isEven(index: Int) = index % 2 == 0

  def separateDigits(number: Long): List[Int] = {
    number.toString.map(_.asDigit).toList
  }

  def sumOfDigits(digits: List[Int]) = {
    if (digits.isEmpty) 0 else separateDigits(digits.mkString.toLong).sum
  }

  def isDivisibleByTenExactly(dividend: Int) = {
    if (dividend < 0) false else dividend % 10 == 0
  }

  private def toLong(number: String) = {
    if (number.length != 11) Left(ValidationError("Credit card numbers must contain 11 digits"))
    else Try(number.toLong).toEither.left.map(toError)
  }

  private val toError: Throwable => ValidationError = {
    case _: NumberFormatException => ValidationError("Credit card numbers must contain only numbers")
  }

}